sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -171,6 +171,63 @@ case class InSubqueryExec(
   }
 }
 
+/**
+ * The physical node of exists-subquery. This is for support use exists in join's on condition,
+ * since some join type we can't pushdown exists condition, we plan it here
+ */
+case class ExistsExec(child: Expression,
+                      subQuery: String,
+                      plan: BaseSubqueryExec,
+                      exprId: ExprId,
+                      private var resultBroadcast: Broadcast[Boolean] = null)
+  extends ExecSubqueryExpression {
+
+  @transient private var result: Boolean = _
+
+  override def dataType: DataType = BooleanType
+  override def children: Seq[Expression] = child :: Nil
+  override def nullable: Boolean = child.nullable
+  override def toString: String = s"EXISTS ${plan.name}"
+  override def withNewPlan(plan: BaseSubqueryExec): ExistsExec = copy(plan = plan)
+
+  override def semanticEquals(other: Expression): Boolean = other match {
+    case in: ExistsExec => child.semanticEquals(in.child) && plan.sameResult(in.plan)
+    case _ => false
+  }
+
+
+  def updateResult(): Unit = {
+    result = !plan.execute().isEmpty()
```

Comment:
seems like this is better to execute a non-correlated EXISTS subquery. Maybe we should update `RewritePredicateSubquery` to only handle correlated EXISTS subquery. @dilipbiswal what do you think?

Reply:
> seems like this is better to execute a non-correlated EXISTS subquery. Maybe we should update `RewritePredicateSubquery` to only handle correlated EXISTS subquery. @dilipbiswal what do you think?

Yeah, wait for his advise.

Reply:
@cloud-fan @AngersZhuuuu Thanks for pinging me. Just for me to understand, since we refer to another pr in this pr.

So we are considering planning the Subqueries appearing inside ON clause as a Join, right ?

Assuming above, so if the query was : 
```
SELECT * FROM T1 JOIN T2 ON T1.C1 = T2.C1 AND T1.C1 EXISTS (SELECT 1 FROM T3 WHERE T1.C1 = T3.C1)
```

We are considering to plan it as : 
```
(T1 LeftSemi T3 ON T1.C1 = T3.C1) Join T2 ON T1.C1 = T2.C2
```

This Looks okay to me for inner joins. I am just not sure about outer joins.. What do you think Wenchen ?

Reply:
@dilipbiswal 
```
SELECT * FROM T1 JOIN T2 ON T1.C1 = T2.C1 AND T1.C1 EXISTS (SELECT 1 FROM T3 WHERE T1.C1 = T3.C1)
```
 Is not correct .

You mean  below ?
```
SELECT * FROM T1 JOIN T2 ON T1.C1 = T2.C1 AND EXISTS (SELECT 1 FROM T3 WHERE T1.C1 = T3.C1)
```

For this type sql we need to change `RewritePredicateSubquery` as cloud-fan said.

Reply:
@AngersZhuuuu You r right. Sorry.. i had written it as IN initially and forgot to adjust to exists :-)

Yeah, we need to change RewritePredicateSubquery which handles correlated subquery rewrites. The only thing i am not sure is about the outer joins.

Reply:
> Yeah, we need to change RewritePredicateSubquery which handles correlated subquery rewrites. The only thing i am not sure is about the outer joins.

Yes, outer join is complex, if we do this, we need to add ent to end test case cover each case to make sure the final plans are as expected.